File: src/game/TRight.ts
import { TSide } from './TSide';
import { TBoard } from './TBoard';
import { GAME_CONSTANTS } from '../constants/GameConstants';

export class TRight extends TSide {
    constructor(hDC: CanvasRenderingContext2D, game: TBoard) {
        super(game, GAME_CONSTANTS.DR);
        
        // Initialize position matching the original
        this.sColor = this.hGame.getCell(0, this.hGame.getW() - 1) - 1;
        this.hGame.setCell(0, this.hGame.getW() - 1, this.playerCode);
        this.score = 1;
        this.check(hDC, 0);
        this.sColor = 0;
        this.change(hDC, 0);
    }

    check(hDC: CanvasRenderingContext2D, forbid: number): number {
        let i: number, j: number, ii: number;
        let backi: boolean, backj: boolean;

        if (this.sColor !== this.fieldColor - 1 && this.sColor !== forbid - 1) {
            this.fieldColor = this.sColor + 1;
        } else {
            return 0;
        }

        i = 0;
        while (i < this.hGame.getH()) {
            backi = false;
            j = this.hGame.getW() - 1;
            while (j >= 0) {
                backj = false;
                ii = i - 1 + (2 * (j % 2));
                
                if (this.hGame.getCell(i, j) & this.playerCode) {
                    this.drawRomb(hDC, i, j, this.fieldColor - 1);
                    
                    if (ii >= 0 && ii < this.hGame.getH()) {
                        if (j + 1 < this.hGame.getW()) {
                            if (this.hGame.getCell(ii, j + 1) === this.fieldColor) {
                                this.hGame.setCell(ii, j + 1, this.playerCode);
                                this.score++;
                                if (ii < i) backi = true;
                                backj = true;
                            }
                        }
                        if (j - 1 >= 0) {
                            if (this.hGame.getCell(ii, j - 1) === this.fieldColor) {
                                this.hGame.setCell(ii, j - 1, this.playerCode);
                                this.score++;
                                if (ii < i) backi = true;
                            }
                        }
                    }
                    
                    if (j + 1 < this.hGame.getW()) {
                        if (this.hGame.getCell(i, j + 1) === this.fieldColor) {
                            this.hGame.setCell(i, j + 1, this.playerCode);
                            this.score++;
                            backj = true;
                        }
                    }
                    if (j - 1 >= 0) {
                        if (this.hGame.getCell(i, j - 1) === this.fieldColor) {
                            this.hGame.setCell(i, j - 1, this.playerCode);
                            this.score++;
                        }
                    }
                    
                    this.hGame.resetCell(i, j, this.playerCode + this.fieldColor);
                }
                
                if (!backj) j--;
                else if (j < this.hGame.getW() - 2) j++;
            }
            if (!backi) i++;
            else if (i) i--;
        }

        this.printScore(hDC);
        
        if (this.score > (this.hGame.getW() * this.hGame.getH() - this.hGame.getW() / 2) / 2) {
            this.animation(hDC);
            return -1;
        } else {
            return 1;
        }
    }

    printScore(hDC: CanvasRenderingContext2D): void {
        hDC.fillStyle = 'rgb(255,255,255)';
        hDC.font = '16px Arial';
        const scoreText = `${this.score}    `;
        hDC.fillText(scoreText, 3 * GAME_CONSTANTS.WIDTH / 4 - GAME_CONSTANTS.RECT_S, GAME_CONSTANTS.HEIGHT - 70);
    }

    selectColor(hDC: CanvasRenderingContext2D, forbid: number): void {
        // AI color selection logic
        // This would implement the computer opponent logic
        this.sColor = Math.floor(Math.random() * GAME_CONSTANTS.NUMBER_COLORS);
        while (this.sColor === forbid - 1) {
            this.sColor = Math.floor(Math.random() * GAME_CONSTANTS.NUMBER_COLORS);
        }
    }

    animation(hDC: CanvasRenderingContext2D): void {
        // Implementation for win animation
        // This would match the original animation logic
    }

    // Override positioning methods for right side
    protected getRectColLeft(): number {
        return 3 * GAME_CONSTANTS.WIDTH / 4 - GAME_CONSTANTS.RECT_S * 2;
    }

    protected getRectColTop(): number {
        return GAME_CONSTANTS.HEIGHT - 93;
    }

    protected getRectColRight(): number {
        return this.getRectColLeft() + GAME_CONSTANTS.RECT_S / 2;
    }

    protected getRectColBottom(): number {
        return this.getRectColTop() + GAME_CONSTANTS.RECT_S / 2;
    }
} 

File: CPP/TRight.cpp
// TRight.cpp: implementation of the TRight class.
//
//////////////////////////////////////////////////////////////////////

#include "stdafx.h"
#include "TRight.h"

//////////////////////////////////////////////////////////////////////
// Construction/Destruction
//////////////////////////////////////////////////////////////////////

TRight::TRight(HDC *hDC,TBoard *Game)
{
	hGame = Game;
	playerCode = 128;

	rectCol.left= 3*WIDTH/4 - RECT_S*2;
	rectCol.top = HEIGHT-93;
	rectCol.right = rectCol.left + RECT_S/2;
	rectCol.bottom = rectCol.top + RECT_S/2 ;

	rect.left = rectCol.left - RECT_S/4;
	rect.top = rectCol.top - RECT_S/4;
	rect.right = rectCol.right + RECT_S/4;
	rect.bottom = rectCol.bottom + RECT_S/4 ;

	SColor = hGame->GetCell(0,hGame->GetW()-1)-1;
	hGame->SetCell(0, hGame->GetW()-1, playerCode);
	Score = 1;
	Check(hDC, 0);
	SColor = 0;
	Change(hDC,0);
}

TRight::~TRight()
{

}
int TRight::Check(HDC *hDC,int forbid)
{
	int i, j, ii;
	bool backi, backj;
	if( SColor != FieldColor-1 && SColor != forbid-1 )
		FieldColor = SColor + 1;
	else return 0;
	i = 0;
	while( i < hGame->GetH())
	{  backi = false;
		j = hGame->GetW() - 1;
		while( j >=0 )
		{  backj = false;
			ii = i-1+(2*(j%2));
			if( hGame->GetCell(i,j) & playerCode )
			{  DrawRomb(hDC, i, j, FieldColor - 1);
				if( ii >= 0 &&  ii < hGame->GetH())
				{	if( j+1 < hGame->GetW() )
						if( hGame->GetCell(ii,j+1) == FieldColor )
						{	hGame->SetCell(ii,j+1, playerCode);
							Score++;
							if( ii < i ) backi = true;
							backj = true; }
					if( j-1 >=0 )
						if( hGame->GetCell(ii,j-1) == FieldColor )
						{  hGame->SetCell(ii,j-1, playerCode);
							Score++;
							if( ii < i ) backi = true;}
				}
				if( j + 1 < hGame->GetW() )
					if( hGame->GetCell(i,j+1) == FieldColor )
					{	hGame->SetCell(i,j+1, playerCode);
						Score++;
						backj = true; }
				if( j - 1 >=0 )
					if( hGame->GetCell(i,j-1) == FieldColor )
					{	hGame->SetCell(i,j-1, playerCode);
						Score++; }
				hGame->ResetCell(i,j, playerCode + FieldColor);
			}
			if( !backj ) j--;
			else if( j < hGame->GetW()-2 ) j++;
		}
		if( !backi ) i++;
		else if ( i ) i--;
	}
	PrintScore(hDC);
	if (Score > (hGame->GetW() * hGame->GetH() - hGame->GetW()/2)/2 )
	{	Animation(hDC);
		return  -1;}
	else  return 1;
}

void TRight::PrintScore(HDC *hDC)
{
	char temp[10];
	SetTextColor( *hDC, RGB(255,255,255 ));
	SetBkColor( *hDC, RGB(0,0,0) );
	wsprintf(temp, "%d%s", Score, "    ");
	TextOut(*hDC,  3*WIDTH/4 - RECT_S, HEIGHT-70, temp, strlen(temp));
}

void TRight::SelectColor(HDC *hDC,int forbid)
{
	int j = hGame->GetW()-1;
	int i = 0;
	while( hGame->GetCell(i, j) & playerCode )
	{ 	i = i + (j%2);
		j--;	}
	if ( i<hGame->GetH()-1 && (!(hGame->GetCell(i, j) & playerCode/2 )) && hGame->GetCell(i, j) != forbid )
	{
	Change(hDC,(SColor>hGame->GetCell(i, j))?(SColor-hGame->GetCell(i, j)-1):(hGame->GetCell(i, j)-SColor-1));
		SColor = hGame->GetCell(i, j) - 1;
	}
	else Change(hDC,1);
}

void TRight::Animation(HDC *hDC)
{
	int i,j;
	for(j = hGame->GetW()-1; j >=0; j--)
		for(i = 0; i < hGame->GetH(); i++)
			if( hGame->GetCell(i, j) )
				DrawRomb(hDC,i, j, FieldColor-1);
}